From 3a66cae1bcb7bdbf71d86a898ed20dde78f1004e Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Mon, 11 Aug 2025 08:59:13 +0200
Subject: [PATCH 912/954] fix(stm32mp13): prevent panic when external
 oscillator is absent

To deactivate the external oscillator, the frequency must be set to 0
in the device tree.
The frequency check for 0 was implemented in clk_stm32_osc_gate_enable()
and clk_stm32_osc_gate_disable(), but missing in
clk_stm32_osc_gate_is_enabled().
Since clk_stm32_osc_gate_is_enabled() is called by clk_enable()
to verify if the clock is actually enabled, this fix prevents a panic
when the oscillator is not present.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git b534e3eb58fd47abaac6562db57f36fe4355bed5 fix(stm32mp13): prevent panic when external oscillator is absent]
Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Change-Id: Ib0c58c4fee29fbf0130264819647386899859f99
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/481608
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
---
 drivers/st/clk/clk-stm32mp13.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/st/clk/clk-stm32mp13.c b/drivers/st/clk/clk-stm32mp13.c
index ecaeef8d92..1f11e97adc 100644
--- a/drivers/st/clk/clk-stm32mp13.c
+++ b/drivers/st/clk/clk-stm32mp13.c
@@ -1805,6 +1805,10 @@ static bool clk_stm32_osc_gate_is_enabled(struct stm32_clk_priv *priv, int id)
 {
 	struct clk_oscillator_data *osc_data = clk_oscillator_get_data(priv, id);
 
+	if (osc_data->frequency == 0UL) {
+		return true;
+	}
+
 	return _clk_stm32_gate_is_enabled(priv, osc_data->gate_id);
 
 }
-- 
2.53.0

